tests/test_experimental.py

```diff
@@ -630,6 +631,290 @@ def multiply(self, X, K):
+        X_prior[zero_weight_indices, :],
+    )
+
+    # Assert AVERAGE variance is reduced everywhere (the sign of ensemble collapse)
+    assert np.mean(posterior_variance_global[zero_weight_indices]) < np.mean(
+        prior_variance[zero_weight_indices]
+    )
+
+    # --- Overall Quality Assertion (MSE) ---
+    # Assert that the localized result is a more plausible estimate
+    mse_localized = np.mean((posterior_mean - true_parameters) ** 2)
+    mse_non_localized = np.mean((posterior_mean_global - true_parameters) ** 2)
+    assert mse_localized < mse_non_localized
+
+
+@pytest.mark.parametrize("seed", list(range(9)))
+def test_distance_based_localization_on_3D_case(seed):
+    # --- 1. System and Assimilation Setup (3D Grid) ---
+    Nx, Ny, Nz = 10, 10, 10  # Dimensions of the 3D parameter grid
+    N_m = Nx * Ny * Nz  # Total number of model parameters (now 1000)
+    N_e = 50  # Ensemble size
+    x_obs, y_obs, z_obs = 5, 5, 5  # Index of the single observation in 3D
+
+    alpha_i = 1
+    obs_error_var = 0.01
+
+    true_parameters = np.zeros(N_m)
+    true_observations = np.array([1.0])
+    C_D = np.array([obs_error_var])
+
+    # --- 2. Generate Initial Ensemble and Predictions ---
+    rng = np.random.default_rng(seed)
+    X_prior = rng.normal(loc=0.0, scale=0.5, size=(N_m, N_e))
+
+    # Convert the 3D observation index to a flat 1D index for slicing
+    flat_obs_index = (z_obs * Nx * Ny) + (y_obs * Nx) + x_obs
+    Y = X_prior[[flat_obs_index], :]
+
+    # --- 3. Construct 3D Localization `rho` ---
+    # A smaller radius is used to ensure the localization effect is clear
+    # and that some weights decay to near-zero within the grid boundaries.
+    localization_radius = 1.5
+
+    # Create 3D coordinate grids
+    zz, yy, xx = np.meshgrid(np.arange(Nz), np.arange(Ny), np.arange(Nx), indexing="ij")
+    # Calculate 3D Euclidean distance from every point to the observation
+    distances_3d = np.sqrt((xx - x_obs) ** 2 + (yy - y_obs) ** 2 + (zz - z_obs) ** 2)
+    distances = distances_3d.flatten()
+
+    rho = np.exp(-0.5 * (distances / localization_radius) ** 2).reshape(-1, 1)
+
+    # --- 4. Run Assimilations ---
+    # Initialize smoothers with the current run's seeded RNG
+    esmda_distance = DistanceESMDA(
+        covariance=C_D, observations=true_observations, alpha=alpha_i, seed=rng
+    )
+    X_posterior = esmda_distance.assimilate(X=X_prior, Y=Y, rho=rho)
+
+    esmda_global = ESMDA(
+        covariance=C_D, observations=true_observations, alpha=alpha_i, seed=rng
+    )
+    X_posterior_global = esmda_global.assimilate(X=X_prior, Y=Y)
+
+    # --- 5. Assertions ---
+    # Find parameters far from the observation where localization weight is near zero
+    zero_weight_indices = np.where(rho < 1e-6)[0]
+    assert len(zero_weight_indices) > 0, (
+        "No distant points found for testing localization."
+    )
+
+    # --- 5a. Assertions for Localized Smoother ---
+    prior_mean = np.mean(X_prior, axis=1)
```

Comment:
This assertions are the same for all tests. I'm wondering if those can be taken to a helper function and thus having just a different setup?

Reply:
Fixed with a separate function doing the asserts